### PR TITLE
feat: reusable schema notifications via gossip_schema crate

### PR DESCRIPTION
**Pull request by domodwyer**
This PR adds the (currently unused) `gossip_schema` crate, that layers the schema-specific messages / logic onto the underlying `gossip` transport crate.

This is designed to make listening for schema changes easy in IOx - initialise a `gossip` instance, initialise the `SchemaRx` and receive a callback for each incoming schema change event :rocket:

A follow-up PR will switch the router to use this crate, removing the bespoke implementation currently in use. All of the logic in this PR is copy/pasted from the current router implementation.

---

* refactor: ColumnsByName::from_iter() (55631a4f8)
      
      Allow a ColumnsByName to be constructed by collecting a set of (name,
      column_schema) tuples.

* feat: implement gossip_schema crate (2e77507f7)
      
      Adds a new gossip_schema crate that provides a high-level interface to
      schema change notifications.
      
      This crate layers schema-specific interfaces over the existing low-level
      gossip crate. Users can obtain best-effort schema change notifications
      by implementing a SchemaEventHandler delegate given to a SchemaRx, or
      efficiently dispatch schema change notifications to listening peers
      using a SchemaTx.
      
      Schema notifications are sent over the Topic::SchemaChanges topic
      (ID=1), which the caller must register as an interest on receiving
      gossip nodes.

### CODE REVIEW

File: gossip_schema/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+//! operation-based CRDT, as the schemas are additive only (append-only sets).
+//!
+//! # Best Effort
+//!
+//! This underlying gossip subsystem is designed to provide best effort delivery
+//! of messages, and therefore best-effort delivery of schema change events,
+//! without any ordering or delivery guarantees.
+//!
+//! This crate does NOT provide any eventual consistency guarantees of schema
+//! data.
```
Comment (domodwyer) [inline gossip_schema/src/lib.rs:63]: This describes how IOx components integrate with this crate, and provides general "what even is this" docs :+1:

File: gossip_schema/src/handle.rs [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+            //    the column ID values reduce to the input set)
+            //
+            // All output messages must contain identical table metadata and be
+            // non-empty.
+
+            // Otherwise validate the successful case.
+            for v in &out {
+                // Invariant: the split frames must be less than or equal to the
+                // desired maximum encoded frame size
+                assert!(v.encoded_len() <= max_frame_bytes);
```
Comment (savage-engineer) [inline gossip_schema/src/handle.rs:273]: Is there a particular reason you don't use `prop_assert!` and `prop_assert_eq!` here?

**Review by savage-engineer — APPROVED** [review]

File: gossip_schema/src/handle.rs (hunk 2 above)
Comment (domodwyer) [reply gossip_schema/src/handle.rs:273]: I find the actual panic output to be helpful when debugging :+1:

File: gossip_schema/src/handle.rs [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            Event::TableCreated(v) => serialise_table_create_frames(v),
+            Event::TableUpdated(v) => {
+                // Split the frame up into N frames, sized as big as the gossip
+                // transport will allow.
+                let mut frames = Vec::with_capacity(1);
+                if !serialise_table_update_frames(v, MAX_USER_PAYLOAD_BYTES, &mut frames) {
+                    warn!("dropping oversized columns in table update");
+                    // Continue sending the columns that were packed
+                    // successfully
+                }
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply gossip_schema/src/handle.rs:82]: Nope - it should be impossible - if we have a single 10KiB column, we're going to have problems elsewhere before gossiping cache optimisation.

Even so, the client must always assume they need to go to the catalog if they see a cache miss!

**Pull request merged**